### ISSUE TITLE
Trimmed videos won't work in firefox html5 video tag

### ISSUE THREAD

**Issue by remes2000** [jcodec/jcodec#465]
I'm trimming my mp4 videos like that:
```
in = NIOUtils.readableChannel(untrimmedVideo);
MP4Util.Movie movie = MP4Util.createRefFullMovie(in, "file://" + untrimmedVideo.getCanonicalPath());
out = NIOUtils.writableChannel(trimmedVideo);
List<MP4Util.Movie> movieSlices = new Cut().cut(movie, Arrays.asList(new Slice(trimRange.getStartPoint(), trimRange.getEndPoint())));
new Strip().strip(movieSlices.get(0).getMoov());
new Flatten().flattenChannel(movieSlices.get(0), out);
```

after that, when i link video tag to trimmed video file everything works fine on chrome and microsoft edge, however only audo is played on firefox. What's worth to mention untrimmed videos (videos that didn't went through jcodec) work perfectly fine on all of the browsers and that's what made me think that it might be a problem with jcodec.

**Comment by jcodec** [comment]
Unfortunately you can not cut and splice videos like that. This is because videos have different kinds of frames, i.e. I,P,B and only I-frames can be decoded on their own, just like images. P and B frames refer to the previously decoded I and P frames. So in order to cut out a slice of video you need to make sure to insert an I-frame at the end of the cut. JCodec doesn't do that and frankly there's no easy way of doing that, as this would require partially transcoding the video to insert those I-frames. The fact that Chrome and Edge play such videos only gives credit to them, engineers went an extra mile to support graceful recovery after an error condition.

**Comment by remes2000** [comment]
So it is not possible to trim mp4 videos correctly with jcodec?

**Comment by jcodec** [comment]
It is possible if you can guarantee that the video will contain I-frames only. For instance you set GOP period to be '1' for x264 encoder. Or you use an I-frame only codec, such as 'ProRes', 'JPEG' or 'PNG'.

**Comment by remes2000** [comment]
In my application users can upload .mp4 videos (that can be trimmed later on) so I don't really have control over what kind of files jcodec will be dealing with. Can I somehow "transpile" those uploaded files to I-frames only format with jcodec?

**Comment by jcodec** [comment]
Yes, you can, but it will be impractical to do so because you would have to 'transcode' portions of the video which would be very slow, since the transcoder is implemented in java and will run on your phone's CPU. Normally phones use a specialized co-processor to encode/decode videos, this specialized hardware takes fraction of power and works a couple of orders of magnitude faster.
The best option in your case would be to only allow your users to cut the videos on I-frame boundaries, you can find out the I-frame boundaries of MP4 file using Jcodec. Additionally, you can transcode uploaded videos using hardware processor into intermediary format that would have a guaranteed I-frame every second or half a second and then use jcodec to cut only at those boundaries.

**Comment by remes2000** [comment]
I run jcodec server-side so I'm not able to take benefits of mobile devices hardware. In my application users can select video to upload and then select start and end point. Then the video is uploaded and trimmed by jcodec. I'm not sure if i understand your advices correctly, so to sum-up what has been said here:
I have two options. The first one is to transcode uploaded video into format that would gurantee I-frame every second and then try to cut it with the same code that i provided above. So It would work like that:
`upload video -> transcode into I-frame every second format -> trim-video -> remove original file`
This approach could be more time consuming so the second solution is to find the nearest I-frame to user specified start-point, and the nearest I-frame to user specified end-point. Then I'll be able to trim the video between those I-frames. In result of that operation trimmed video could be a little too short or a little too long (relative to what user wanted to achieve), but won't be damaged as it is now.

**Comment by jcodec** [comment]
Yes, you can use jcodec to find out I-frame map stored in the MP4 header. Then use this map to cut the videos. You don't have to cut both sides to I-frame, only the right-side has to start with an I-frame. Correct, the videos will be a little too short or a little too long. To minimize negative effect you can make sure I-frames are available quite often  (say every 0.5s) for this you will need to transcode the video first.

Additionally, you can check if your browsers support MP4 edit lists. If they do, you can simply add an edit into MP4 header and players will automatically skip the portion of the video specified outside the edit. JCodec will allow you to create an edit list inside the MP4 header. You can combine both solutions together, i.e. first you take out approximately the correct amount of video with respect to I-frames, and then you introduce an edit into MP4 header that takes out the rest, so the videos are always cut precisely to the interval.

**Comment by remes2000** [comment]
I think I get it. Here is what my trimming UI looks like:
![github_1](https://user-images.githubusercontent.com/26173223/124780005-8bba2580-df42-11eb-9a48-92ac849e4427.PNG)

So in fact user selects only that part of video that he wants to be presented later in application. In that case I only have to make sure that it starts with an I-frame? Since P and B frames refer to previous frame, I have to make sure that first frame in video is independent and it doesn't point at non-existing frame. Another words in my ui left pointer has to point at I frame (not P or B) while position of right pointer doesn't matter since it is the last frame it can only points at a frame that is included in file or be independent. (Please correct me if my understanding is wrong).
Since step of slider in my UI is set to one second, I-frame in every second will give me enough trimming accuracy. I believe I understand what should I do and how, but I'm having trouble with finding some kind of documentation or at least code examples of how to use jcodec. Could you provide me any tips/code snippets/documentation pages of how to accomplish transcoding .mp4 file to I-frame every second format and how to find I-frames in mp4 file?

**Comment by jcodec** [comment]
I would use export the list I-frames to the UI and let the left slider only land where there's an I-frame. That way the UI will always be consistent with the backend trimming operation.
To the full list of I-frames:
```
MP4Demuxer demuxer = .... // create MP4 demuxer
DemuxerTrack videoTrack = demuxer.getVideoTracks().get(0);
DemuxerTrackMeta meta = videoTrack.getMeta();
int[] seekFrames = meta.getSeekFrames();
// seekFrames will literally contain frame indexes of I-frames
double fps = meta.getTotalFrames() / meta.getTotalDuration();
for (int frameNo : seekFrames) {
  // This would be roughly ok
  double frameSec = frameNo*fps;
  // To get each frame pts for sure need to go through the file with demuxer, i.e.
  // ((SeekableDemuxerTrack)videoTrack).gotoFrame(frameNo);
  // Packet pkt = videoTrack.nextFrame();
  // double frameSec = pkt.getPtsD();
}
```

**Comment by remes2000** [comment]
Thanks, you helped me a lot. I'm gonna try to implement this solution now. These few messages made me understand where the problem is. 

I won't be able to calibrate slider to land only on I-frames since user manipulate unuploaded video and I would have to somehow get list o I-frames from file on frontend of my application. That might be hard to achieve in javascript. So I'll try to find first I-frame before start point and trim from that moment.

**Comment by jcodec** [comment]
sure, if you ever decide to make this I-frame map available on the UI, there are multiple options.
You can find the header portion using JS and upload it to the server, header portion only really takes a couple of kilobytes for small videos usually. The browser will ask a special permission from the user in that case.
You can even use the entire JCodec cross-compiled to javascript to avoid uploading alltogether, we tried JS Jcodec before, works fine.
